syntax = "proto3";

package models;

message Company {
    string id = 1;
    string name = 2;
    string description = 3;
    string email = 4;
    string password = 5;
    bool is_verified = 6;
    string verified_at = 7;
    string created_at = 8;
    string updated_at = 9;
    repeated CompanyToken tokens = 10;
    repeated Product products = 11;
}

message CompanyToken {
    string id = 1;
    string company_id = 2;
    string token = 3;
    string created_at = 4; // ISO 8601 format
    string expired_at = 5; // ISO 8601 format
}

message User {
    string id = 1;
    string first_name = 2;
    string last_name = 3;
    string password = 4;
    string email = 5;
    bool is_verified = 6;
    string verified_at = 7;
    string created_at = 8;
    string updated_at = 9;
    repeated UserToken tokens = 10;
    repeated Order orders = 11;
    repeated Comment comments = 12;
    repeated Cart carts = 13;
}

message UserToken {
    string id = 1;
    string user_id = 2;
    string token = 3;
    string created_at = 4; // ISO 8601 format
    string expired_at = 5; // ISO 8601 format
}

message Comment {
    string id = 1;
    string user_id = 2;
    string product_id = 3;
    string comment_text = 4;
    float rating = 5;
    string created_at = 6; // ISO 8601 format
    string updated_at = 7; // ISO 8601 format
}

message Email {
    string id = 1;
    string recipient_email = 2;
    string subject = 3;
    string sent_at = 4; // ISO 8601 format
    string status = 5;
    repeated EmailContent contents = 6;
}

message EmailContent {
    string id = 1;
    string content = 2;
    string email_id = 3;
}

message Product {
    string id = 1;
    string name = 2;
    string description = 3;
    float price = 4;
    string category = 5;
    string company_id = 6;
    float product_ranking = 7;
    string created_at = 8;
    string updated_at = 9;
    repeated Comment comments = 10;
    repeated Order orders = 11;
    StockLevel stock = 12;
    repeated InventoryTransaction transactions = 13;
}

message Cart {
    string id = 1;
    string user_id = 2;
    string status = 3;
    string created_at = 4;
    string updated_at = 5;
    float total_price = 6;
    repeated Order orders = 7;
}

message Order {
    string id = 1;
    string cart_id = 2;
    string user_id = 3;
    string product_id = 4;
    float product_price = 5;
    int32 quantity = 6;
    float total_price = 7;
    string order_date = 8;
    string delivery_date = 9;
    bool delivered = 10;
    string status = 11;
    string return_back = 12;
    string delivery_address = 13;
    string created_at = 14;
    string updated_at = 15;
}

message InventoryTransaction {
    string id = 1;
    string stock_id = 2;
    string product_id = 3;
    int32 quantity = 4;
    string timestamp = 5;
    string operation = 6;
    Product product = 7 ;
    StockLevel stock = 8;  
}

message StockLevel {
    string id = 1;
    string product_id = 2;
    int32 quantity = 3;
    string updated_at = 4;
    string created_at = 5;
    repeated InventoryTransaction transactions = 6;
}

message ProductWithStock {
    string id = 1;
    string name = 2;
    string description = 3;
    float price = 4;
    string category = 5;
    string company_id = 6;
    float product_ranking = 7;
    string created_at = 8;
    string updated_at = 9;
    string stock = 10;
}